Framework/Algorithms/src/AverageSpectrumBackground.cpp

```diff
@@ -0,0 +1,151 @@
+}
+
+/// Algorithm's summary for use in the GUI and help. @see Algorithm::summary
+const std::string AverageSpectrumBackground::summary() const { return ""; }
+
+// Register the algorithm into the AlgorithmFactory
+DECLARE_ALGORITHM(AverageSpectrumBackground)
+
+/** Sums spectra bin by bin in the given range using the child algorithm
+ * GroupDetectors. Note that KeepUngroupedSpectra is set to false.
+ * @param inputWS:: The Workspace2D to take as input
+ * @param indexList:: An array of workspace indices to combine
+ * @return :: A workspace containing a single spectra and DectectorGroups
+ */
+API::MatrixWorkspace_sptr AverageSpectrumBackground::groupBackgroundDetectors(
+    API::MatrixWorkspace_sptr inputWS, const std::vector<size_t> indexList) {
+
+  auto alg = this->createChildAlgorithm("GroupDetectors");
+  alg->setProperty("InputWorkspace", inputWS);
+  alg->setProperty("WorkspaceIndexList", indexList);
+  alg->setProperty("KeepUngroupedSpectra", false);
+  alg->execute();
+  return alg->getProperty("OutputWorkspace");
+}
+
+std::vector<size_t> AverageSpectrumBackground::getSpectraFromRange(
+    const std::vector<size_t> range) {
+  const auto start = range[0];
+  const auto end = range[1];
+  std::vector<size_t> spectra;
+  for (auto itr = start; itr < end; ++itr) {
+    spectra.push_back(itr);
+  }
+  return spectra;
+}
+
+//----------------------------------------------------------------------------------------------
+/** Initialize the algorithm's properties.
+ */
+void AverageSpectrumBackground::init() {
+
+  // Input workspace
+  declareProperty(
+      Kernel::make_unique<API::WorkspaceProperty<API::MatrixWorkspace>>(
+          "InputWorkspace", "", Kernel::Direction::Input,
+          boost::make_shared<API::CommonBinsValidator>()),
+      "An input workspace.");
+
+  // bottom background range
+  declareProperty(Kernel::make_unique<Kernel::ArrayProperty<size_t>>(
+                      "BottomBackgroundRange", std::vector<size_t>()),
+                  "A list of the bottom background ranges.");
+  // top background range
+  declareProperty(Kernel::make_unique<Kernel::ArrayProperty<size_t>>(
+                      "TopBackgroundRange", std::vector<size_t>()),
+                  "A list of the top background ranges.");
+
+  // Output workspace
+  declareProperty(Kernel::make_unique<API::WorkspaceProperty<>>(
+                      "OutputWorkspace", "", Kernel::Direction::Output),
+                  "A Workspace with the background removed.");
+}
+
+//----------------------------------------------------------------------------------------------
+/** Execute the algorithm.
+ */
+void AverageSpectrumBackground::exec() {
+  API::MatrixWorkspace_sptr inputWS = getProperty("InputWorkspace");
+  std::vector<size_t> bottomBgdRange = getProperty("BottomBackgroundRange");
+  std::vector<size_t> topBgdRange = getProperty("TopBackgroundRange");
+
+  if (bottomBgdRange.empty() && topBgdRange.empty())
```

Comment:
I think some of this validation might not be necessary if we can use IndexProperty instead of an array. The rest could be done in a `validateInputs` method e.g. see [here](https://github.com/mantidproject/mantid/blob/master/Framework/Algorithms/src/CreateUserDefinedBackground.cpp#L236).